src/main/scala/com/scylladb/migrator/writers/DynamoStreamReplication.scala

```diff
@@ -62,54 +54,39 @@ object DynamoStreamReplication {
 
           newMap.putAll(rec.getDynamodb.getKeys)
 
-          val row =
-            Row.fromSeq {
-              typeConverter
-                .readInternalRow(ItemUtils.toItem(newMap))
-                .toSeq(inferredSchema)
-                .map {
-                  case s: UTF8String => s.toString
-                  case x             => x
-                } ++
-                Seq(
-                  rec.getEventName match {
-                    case "INSERT" | "MODIFY" => ColumnSchema.PutOperation
-                    case "REMOVE"            => ColumnSchema.DeleteOperation
-                  }
-                )
+          val operationType =
+            rec.getEventName match {
+              case "INSERT" | "MODIFY" => putOperation
+              case "REMOVE"            => deleteOperation
             }
-
-          Some(row)
+          newMap.put(operationTypeColumn, operationType)
+          Some(newMap)
 
         case _ => None
       }
       .foreachRDD { msgs =>
-        val df = spark
-          .createDataFrame(
-            msgs
-              .collect {
-                case Some(row) => row
-              },
-            inferredSchema.add(ColumnSchema.OperationTypeColumn, BooleanType)
-          )
-          .repartition(partitionColumns: _*)
+        val rdd = msgs.collect {
+          case Some(item) => (new Text(), new DynamoDBItemWritable(item))
+        }
+        // FIXME re-partition by columns? msgs.repartition(???)
```

Comment:
In the previous implementation, we used to `repartition` the items by [`partitionColumns`](https://github.com/scylladb/scylla-migrator/pull/105/files#diff-cbd503faf5f6999879a56b60c9beb2ef02dccb9d4c013bc6a8128c9f2ca3dffaL37). I am not sure why this was useful. Repartitioning by columns does not seem to be natively supported on `RDD`s (only on `DataFrame`s), so I didn’t try to achieve something “equivalent” at the RDD level.

Reply:
it's likely due to shuffle

the problem is that when you start reading partitions in order you will saturate only single cpu that usually owns sequential tokens (PKs)
if you want to maximize parallelism you need to have various random tokens being read in parallel - this makes ALL cpus busy due to how partitions(token ranges) are divided between them

so if we can shuffle the tasks created on top of RDD then we would achieve the same, is this doable? (I know spark 3 tries to improve this even more to avoid explicit shuffle need, but RDDs might need it, especially if their partitioner generates tasks in order)

Reply:
https://docs.aws.amazon.com/prescriptive-guidance/latest/tuning-aws-glue-for-apache-spark/parallelize-tasks.html explains
so repartition is needed to have more partitions, so basically 2-3x more than cpus
so again it's about parallelism/size of the tasks that should ideally be shuffled

Reply:
Thank you for the pointer.

What was previously done is what is described on that page:

> However, when writing a partitioned table, this can lead to an explosion of files (each partition can potentially generate a file into each table partition). To avoid this, you can repartition your DataFrame by column. This uses the table partition columns so the data is organized before writing. You can specify a higher number of partitions without getting small files on the table partitions. However, be careful to avoid data skew, in which some partition values end up with most of the data and delay the completion of the task.

However, we cannot repartion by column anymore since we don’t have a schema anymore. I can try the “2-3× more than CPUs” rule, though.

Reply:
this eventually should be a tunable, after all this in the end will be the number of map reduce tasks on RDD side
which will impact the whole speed and performance of solution

@pdbossman we recently figure out a good split rate , do you have the maths behind it handy? can you share that tuning (related to dynamo DB side?)

Reply:
ok, I think me and Pat came to this re num partitions:
https://github.com/audienceproject/spark-dynamodb/blob/816c6e6d8a250a2d8b700761b94a198a872a7ea6/src/main/scala/com/audienceproject/spark/dynamodb/connector/TableConnector.scala#L64

Reply:
but I just noticed it's for stream ... do we even care?
but maybe we do ... if there is huge amount of writes in original table, we need to scale out for them and be ready eventually

Reply:
I am not sure how important this is, to be honest. I did it because the existing implementation does something similar:

https://github.com/scylladb/scylla-migrator/blob/a7d43e0c054be32631974b2ad7d595f806f35267/migrator/src/main/scala/com/scylladb/migrator/writers/DynamoStreamReplication.scala#L95

Reply:
I expect it's about scalability for high throughput - if streaming will do lots of writes

Reply:
> ok, I think me and Pat came to this re num partitions: https://github.com/audienceproject/spark-dynamodb/blob/816c6e6d8a250a2d8b700761b94a198a872a7ea6/src/main/scala/com/audienceproject/spark/dynamodb/connector/TableConnector.scala#L64

Here is the relevant piece of code from the link:

~~~ scala
  val numPartitions = parameters.get("readpartitions").map(_.toInt).getOrElse({
      val sizeBased = (tableSize / maxPartitionBytes).toInt max 1
      val remainder = sizeBased % parallelism
      if (remainder > 0) sizeBased + (parallelism - remainder)
      else sizeBased
  })
~~~

One challenge to implement it in the context of streamed changes is that there is no fast and accurate way to get the size of an RDD. We would also have to introduce more settings to get e.g. the `maxPartitionBytes` value.